File: skills/install-slop-guard/scripts/detect.mjs
// Language detection shared by the installer and the scanner.
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { dirname, extname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export const skillRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..");
export const registry = JSON.parse(readFileSync(join(skillRoot, "languages.json"), "utf8"));

const FILE_LIMIT = 20000;

function sourceExtensions(root, excluded) {
  const found = [];
  const queue = [root];
  while (queue.length > 0 && found.length < FILE_LIMIT) {
    const directory = queue.shift();
    let entries;
    try {
      entries = readdirSync(directory, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!excluded.has(entry.name)) queue.push(join(directory, entry.name));
      } else if (found.length < FILE_LIMIT) {
        found.push(extname(entry.name));
      }
    }
  }
  return found;
}

/** Report which registered languages a directory appears to contain, most source files first. */
export function detect(root) {
  const excluded = new Set(registry.exclude);
  const extensions = sourceExtensions(root, excluded);
  return registry.languages
    .map((language) => {
      const markers = language.markers.filter((marker) => existsSync(join(root, marker)));
      const sources = extensions.filter((extension) => language.extensions.includes(extension)).length;
      return { language, markers, sources, detected: markers.length > 0 || sources > 0 };
    })
    .filter((result) => result.detected)
    .sort((left, right) => right.sources - left.sources);
}

/** Look up one language by its id, or throw with the ids this skill knows. */
export function language(id) {
  const found = registry.languages.find((entry) => entry.id === id);
  if (found === undefined) {
    const known = registry.languages.map((entry) => entry.id).join(", ");
    throw new Error(`Unknown language ${id}. This skill supports: ${known}.`);
  }
  return found;
}

File: skills/install-slop-guard/scripts/detect.py
"""Language detection shared by the installer and the scanner."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

SKILL_ROOT = Path(__file__).resolve().parent.parent
REGISTRY = json.loads((SKILL_ROOT / "languages.json").read_text(encoding="utf-8"))
FILE_LIMIT = 20000


@dataclass(frozen=True, slots=True)
class Detection:
    """One language the repository appears to contain."""

    language: dict
    markers: tuple[str, ...]
    sources: int


def _source_extensions(root: Path, excluded: set[str]) -> list[str]:
    found: list[str] = []
    queue = [root]
    while queue and len(found) < FILE_LIMIT:
        directory = queue.pop(0)
        try:
            entries = sorted(directory.iterdir())
        except OSError:
            continue
        for entry in entries:
            if entry.is_dir():
                if entry.name not in excluded:
                    queue.append(entry)
            elif len(found) < FILE_LIMIT:
                found.append(entry.suffix)
    return found


def detect(root: Path) -> list[Detection]:
    """Report which registered languages a directory appears to contain."""
    excluded = set(REGISTRY["exclude"])
    extensions = _source_extensions(root, excluded)
    results: list[Detection] = []
    for language in REGISTRY["languages"]:
        markers = tuple(name for name in language["markers"] if (root / name).exists())
        sources = sum(1 for suffix in extensions if suffix in language["extensions"])
        if markers or sources:
            results.append(Detection(language=language, markers=markers, sources=sources))
    return sorted(results, key=lambda result: result.sources, reverse=True)


def language(identifier: str) -> dict:
    """Look up one language by its id, or raise with the ids this skill knows."""
    for entry in REGISTRY["languages"]:
        if entry["id"] == identifier:
            return entry
    known = ", ".join(entry["id"] for entry in REGISTRY["languages"])
    raise KeyError(f"Unknown language {identifier}. This skill supports: {known}.")
